### PR TITLE
ICU-20099 Implementing all current Java ListFormatter proposals.

### PR DESCRIPTION
**Pull request by sffc**
This includes both of the current Java ListFormatter proposals.  The PR also includes some refactoring on SimpleFormatterImpl (ICU-20406).

##### Checklist

- [x] Issue filed: https://unicode-org.atlassian.net/browse/ICU-20099
- [x] Updated PR title and link in previous line to include Issue number
- [x] Issue accepted
- [x] Tests included
- [x] Documentation is changed or added

### CODE REVIEW

**Comment by sffc** [comment]
Adding Markus as an optional reviewer and Yoshito as the primary reviewer.

File: icu4j/main/classes/core/src/com/ibm/icu/text/ListFormatter.java [hunk 1]
```diff
@@ -98,6 +105,241 @@ public String getName() {
...
+        public static final SpanField LIST_SPAN = new SpanField("list-span");
+
+        private SpanField(String name) {
+            super(name);
+        }
+
+        /**
+         * serizalization method resolve instances to the constant
+         * ListFormatter.SpanField values
+         * @draft ICU 64
```
Comment (yumaoka) [inline icu4j/main/classes/core/src/com/ibm/icu/text/ListFormatter.java]: Please change this to `@internal`

File: icu4j/main/classes/core/src/com/ibm/icu/text/ListFormatter.java [hunk 2]
```diff
@@ -177,6 +425,32 @@ public static ListFormatter getInstance(ULocale locale, Style style) {
         return cache.get(locale, style.getName());
     }
 
+    /**
+     * Create a list formatter that is appropriate for a locale.
+     *
+     * @param locale
+     *            the locale in question.
+     * @return ListFormatter
+     * @draft ICU 67
```
Comment (yumaoka) [inline icu4j/main/classes/core/src/com/ibm/icu/text/ListFormatter.java]: This one should be `@stable ICU 50`

File: icu4j/main/classes/core/src/com/ibm/icu/text/ListFormatter.java [hunk 3]
```diff
@@ -177,6 +425,32 @@ public static ListFormatter getInstance(ULocale locale, Style style) {
...
+      return getInstance(locale, Style.STANDARD);
+    }
+
+    /**
+     * Create a list formatter that is appropriate for a locale.
+     *
+     * @param locale
+     *            the locale in question.
+     * @return ListFormatter
+     * @draft ICU 67
```
Comment (yumaoka) [inline icu4j/main/classes/core/src/com/ibm/icu/text/ListFormatter.java]: This one should be `@stable ICU 50`

File: icu4j/main/classes/core/src/com/ibm/icu/text/ListFormatter.java [hunk 4]
```diff
@@ -348,4 +661,46 @@ private static ListFormatter load(ULocale ulocale, String style) {
...
+        switch (type) {
+            case AND:
+                switch (width) {
+                    case WIDE:
+                        return "standard";
+                    case SHORT:
+                        return "standard-short";
+                    case NARROW:
+                        return "standard-narrow";
+                    default:
```
Comment (yumaoka) [inline icu4j/main/classes/core/src/com/ibm/icu/text/ListFormatter.java]: Can we remove `default:` ?
When we `Width` enum is added, we want Java compiler to warn it.

File: icu4j/main/classes/core/src/com/ibm/icu/text/ListFormatter.java [hunk 5]
```diff
@@ -98,6 +105,241 @@ public String getName() {
 
     }
 
+    /**
+     * Type of meaning expressed by the list.
+     *
+     * @draft ICU 67
+     * @provisional This API might change or be removed in a future release.
+     */
+    public enum ListFormatterType {
```
[the remark this answers is not in the document]
Comment (sffc) [reply icu4j/main/classes/core/src/com/ibm/icu/text/ListFormatter.java]: Done.

File: icu4j/main/classes/core/src/com/ibm/icu/text/ListFormatter.java [hunk 6]
```diff
@@ -98,6 +105,241 @@ public String getName() {
...
+        UNIT
+    };
+
+    /**
+     * Verbosity level of the list patterns.
+     *
+     * @draft ICU 67
+     * @provisional This API might change or be removed in a future release.
+     */
+    public enum ListFormatterWidth {
```
[the remark this answers is not in the document]
Comment (sffc) [reply icu4j/main/classes/core/src/com/ibm/icu/text/ListFormatter.java]: Done.

File: icu4j/main/classes/core/src/com/ibm/icu/text/ListFormatter.java (hunk 1 above)
Comment (sffc) [reply icu4j/main/classes/core/src/com/ibm/icu/text/ListFormatter.java]: Done, although I seem to remember saying that we would follow precedent and put these serialization methods on track to stable like we do in other cases.  I don't have an opinion either way.

File: icu4j/main/classes/core/src/com/ibm/icu/text/ListFormatter.java (hunk 3 above)
Comment (sffc) [reply icu4j/main/classes/core/src/com/ibm/icu/text/ListFormatter.java]: Done.  Don't know why it changed.

File: icu4j/main/classes/core/src/com/ibm/icu/text/ListFormatter.java (hunk 2 above)
Comment (sffc) [reply icu4j/main/classes/core/src/com/ibm/icu/text/ListFormatter.java]: Done.  Don't know why it changed.

File: icu4j/main/classes/core/src/com/ibm/icu/text/ListFormatter.java (hunk 4 above)
Comment (sffc) [reply icu4j/main/classes/core/src/com/ibm/icu/text/ListFormatter.java]: Done.

**Comment by sffc** [comment]
@yumaoka  PTAL

**Comment by sffc** [comment]
Squashed.  PTAL

**Comment by sffc** [comment]
Ping (I squashed and need another stamp)

**Comment by sffc** [comment]
@yumaoka Ping

**Pull request merged**